app/services/katello/pulp3/content.rb

```diff
@@ -5,17 +5,29 @@ class Content
       extend Katello::Abstract::Pulp::Content
       class << self
         def create_upload(size = 0, checksum = nil, content_type = nil)
-          content_unit_href = nil
           if checksum && content_type
             content_backend_service = SmartProxy.pulp_master.content_service(content_type)
             content_list = content_backend_service.content_api.list("digest": checksum)
-            content_unit_href = content_list.results.first.pulp_href unless content_list.results.empty?
-            return {"content_unit_href" => content_unit_href} if content_unit_href
+            content_href = content_href_by_checksum(content_type, content_list.results, checksum, content_backend_service.content_api)
+            return content_href if content_href
           end
           upload_href = uploads_api.create(upload_class.new(size: size)).pulp_href
           {"upload_id" => upload_href.split("/").last}
         end
 
+        def content_href_by_checksum(content_type, content_list_results, checksum, content_api)
+          if content_type == FileUnit::CONTENT_TYPE
+            content_unit_href = content_list_results.first.pulp_href unless content_list_results.empty?
+            return {"content_unit_href" => content_unit_href} if content_unit_href
+          else
+            content_list_results.each do |package|
+              package_details = content_api.read(package.pulp_href)
+              return {"content_unit_href" => package.pulp_href} if package_details.pkg_id == checksum
```

Comment:
Is this with or without the changes that the pulp team was going to add?

I don't quite get the logic here, its fetching all rpms with the same checksum, passing that to this fucntion, then this is looping over the results and fetching each full unit (is that required?), but then its comparing the pkg_id to the expected checksum, what is pkg_id here?

Reply:
Currently there is no "digest" option for other plugins, so what you get is a list of all packages. There is a request for pulp_rpm to add the ability to filter a list by checksum, or an equivalent, but it doesn't exist at the moment.

For large package lists, this seems like a workaround that will become cumbersome. I couldn't think of another way to do this. Hopefully the pulp team will add this and we can refactor to take advantage of it.

For each ContentPackage object returned, the `pkg_id` is the checksum (according the RpmPackage documentation).

Reply:
so i guess the call to `content_backend_service.content_api.list("digest": checksum)` just returns all rpms?    I'm not sure that'll work properly since its not paged.  I'd kinda be in favor of just letting this fail for duplicate uploads until the pulp team adds digest support.  It would simplify it a bit too